src/samplers/v2/sampling_state.js

```diff
@@ -0,0 +1,143 @@
+// @flow
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+import { DEBUG_MASK, FIREHOSE_MASK, SAMPLED_MASK } from '../../constants';
+import Span from '../../span';
+import SpanContext from '../../span_context';
+
+export const WARN_NOT_FINALIZABLE = 'This sampling state is not finalizable.';
+export const WARN_SHOULD_NOT_ENABLE_FINALIZABLE = 'Finalizable can only be changed to false';
+export const WARN_CANNOT_REVERT_FINALIZABLE = 'Setting finalizable to false cannot be reverted';
+export const WARN_CANNOT_REVERT_FINAL = 'The final state cannot be reverted.';
+// export const WARN_CANNOT_UNSAMPLE = 'The final state cannot be reverted.';
+// export const WARN_CANNOT_REMOVE_DEBUG = 'The debug flag cannot be reverted.';
+// export const WARN_CANNOT_REMOVE_FIREHOSE = 'The firehose flag cannot be reverted.';
+
+export default class SamplingState {
+  _extendedState: { [string]: any } = {};
+  _flags: number = 0;
+  _isFinal: boolean = false;
+  _isFinalizable: boolean = true;
```

Comment:
Naming looks a bit odd to me. The traditional naming is

var _final: bool
function isFinal(): bool
function setFinal(value: bool)

Reply:
SGTM